File: crates/zmk-layout-python/examples/recreate_tailorkey.py
#!/usr/bin/env python3
"""Recreate the TailorKey sample keymap using code and small data modules."""

from __future__ import annotations

import json
import sys
from pathlib import Path
from typing import Any, Dict, List

import zmk_layout

ROOT = Path(__file__).resolve().parents[3]
sys.path.insert(0, str(Path(__file__).parent))

from tailorkey_layers import base, layer_names, overrides  # noqa: E402
from tailorkey_behaviors import hold_taps, macros  # noqa: E402
from tailorkey_combos import combos, input_listeners, metadata  # noqa: E402


def build_layers() -> List[Dict[str, Any]]:
    layers = [{"name": layer_names[0], "bindings": base}]
    for name in layer_names[1:]:
        bindings = ["&trans"] * len(base)
        for idx, val in overrides.get(name, {}).items():
            bindings[idx - 1] = val
        layers.append({"name": name, "bindings": bindings})
    return layers


def build_macros() -> List[Dict[str, Any]]:
    result = []
    for macro in macros:
        cells = len(macro.get("params", []))
        result.append(
            {
                "name": macro["name"],
                "description": macro.get("description", ""),
                "bindings": macro.get("bindings", []),
                "wait_ms": macro.get("wait_ms"),
                "tap_ms": macro.get("tap_ms"),
                "binding_cells": cells or None,
                "compatible": (
                    "zmk,behavior-macro-one-param" if cells else "zmk,behavior-macro"
                ),
            }
        )
    return result


def fmt_prop(value: Any) -> str | None:
    if value is None:
        return None
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, (int, float)):
        return f"< {value} >"
    if isinstance(value, list):
        return "< " + " ".join(str(v) for v in value) + " >"
    if isinstance(value, str):
        return f'"{value}"'
    return str(value)


def build_behaviors() -> List[Dict[str, Any]]:
    result = []
    for ht in hold_taps:
        props = {
            "tapping-term-ms": fmt_prop(ht.get("tapping_term_ms")),
            "quick-tap-ms": fmt_prop(ht.get("quick_tap_ms")),
            "require-prior-idle-ms": fmt_prop(ht.get("require_prior_idle_ms")),
        }
        positions = ht.get("hold_trigger_key_positions") or []
        if positions:
            props["hold-trigger-key-positions"] = fmt_prop(positions)
        if "hold_trigger_on_release" in ht:
            props["hold-trigger-on-release"] = fmt_prop(ht.get("hold_trigger_on_release"))
        if "flavor" in ht and ht["flavor"] is not None:
            props["flavor"] = fmt_prop(ht["flavor"])
        # drop None values
        props = {k: v for k, v in props.items() if v is not None}
        result.append(
            {
                "name": ht["name"],
                "description": ht.get("description", ""),
                "compatible": "zmk,behavior-hold-tap",
                "binding_cells": 2,
                "bindings": ht.get("bindings", []),
                "properties": props,
            }
        )
    return result


def build_standard_layout() -> Dict[str, Any]:
    return {
        "layers": build_layers(),
        "combos": combos,
        "behaviors": build_behaviors(),
        "macros": build_macros(),
        "input_listeners": input_listeners,
        "metadata": metadata,
    }


def main() -> None:
    layout = zmk_layout.Layout()
    standard = build_standard_layout()
    json_payload = json.dumps(standard)

    layout.parse_json(json_payload, "examples/moergo_glove80.j2")

    output_path = ROOT / "out" / "tailorkey_from_json.keymap"
    output_path.parent.mkdir(parents=True, exist_ok=True)
    layout.save_dts(str(output_path))
    print(f"Recreated TailorKey -> {output_path}")


if __name__ == "__main__":
    main()

File: examples/recreate_tailorkey.lua
-- Recreate the TailorKey sample keymap from code (no external JSON reads).
-- Data lives in small modules and we build the standard layout table, encode
-- it to JSON, and render via the Glove80 template.
--
-- Run:
--   zmk-layout keymap lua \
--     --script examples/recreate_tailorkey.lua \
--     --layout /dev/null \
--     --output out/tailorkey_from_json.keymap

local layers_data = dofile("examples/tailorkey_layers.lua")
local behaviors_data = dofile("examples/tailorkey_behaviors.lua")
local combos_data = dofile("examples/tailorkey_combos.lua")

-- JSON encoder (minimal, handles tables, strings, numbers, bools, nil).
local function escape(str)
    return (str:gsub("\\", "\\\\")
        :gsub("\"", "\\\"")
        :gsub("\b", "\\b")
        :gsub("\f", "\\f")
        :gsub("\n", "\\n")
        :gsub("\r", "\\r")
        :gsub("\t", "\\t"))
end

local function is_array(tbl)
    local max_idx = 0
    local count = 0
    for k, _ in pairs(tbl) do
        if type(k) ~= "number" then
            return false
        end
        if k > max_idx then
            max_idx = k
        end
        count = count + 1
    end
    if max_idx == 0 then
        return true, 0
    end
    for i = 1, max_idx do
        if tbl[i] == nil then
            return false
        end
    end
    return true, max_idx
end

local function encode_json(value)
    local t = type(value)
    if t == "string" then
        return "\"" .. escape(value) .. "\""
    elseif t == "number" or t == "boolean" then
        return tostring(value)
    elseif t == "nil" then
        return "null"
    elseif t == "table" then
        local array, max_idx = is_array(value)
        if array then
            local parts = {}
            for i = 1, max_idx do
                parts[i] = encode_json(value[i])
            end
            return "[" .. table.concat(parts, ",") .. "]"
        else
            local parts = {}
            for k, v in pairs(value) do
                parts[#parts + 1] = "\"" .. escape(k) .. "\":" .. encode_json(v)
            end
            return "{" .. table.concat(parts, ",") .. "}"
        end
    else
        error("unsupported json type: " .. t)
    end
end

local function build_layers()
    local built = {}
    local base = layers_data.base
    built[1] = { name = layers_data.layer_names[1], bindings = base }
    for i = 2, #layers_data.layer_names do
        local name = layers_data.layer_names[i]
        local overrides = layers_data.overrides[name] or {}
        local bindings = {}
        for idx = 1, #base do
            bindings[idx] = "&trans"
        end
        for idx, val in pairs(overrides) do
            bindings[idx] = val
        end
        built[#built + 1] = { name = name, bindings = bindings }
    end
    return built
end

local function build_macros()
    local result = {}
    for _, macro in ipairs(behaviors_data.macros) do
        local cells = macro.params and #macro.params or 0
        result[#result + 1] = {
            name = macro.name,
            description = macro.description or "",
            bindings = macro.bindings,
            wait_ms = macro.wait_ms,
            tap_ms = macro.tap_ms,
            binding_cells = cells > 0 and cells or nil,
            compatible = cells > 0 and "zmk,behavior-macro-one-param" or "zmk,behavior-macro",
        }
    end
    return result
end

local function format_property(value)
    if value == nil then
        return nil
    end
    local t = type(value)
    if t == "number" then
        return string.format("< %s >", value)
    elseif t == "boolean" then
        return value and "true" or "false"
    elseif t == "table" then
        local parts = {}
        for i, v in ipairs(value) do
            parts[i] = tostring(v)
        end
        return "< " .. table.concat(parts, " ") .. " >"
    elseif t == "string" then
        return string.format("\"%s\"", value)
    end
    return tostring(value)
end

local function build_behaviors()
    local result = {}
    for _, ht in ipairs(behaviors_data.hold_taps) do
        local props = {}
        props["tapping-term-ms"] = format_property(ht.tapping_term_ms)
        props["quick-tap-ms"] = format_property(ht.quick_tap_ms)
        props["require-prior-idle-ms"] = format_property(ht.require_prior_idle_ms)
        if ht.hold_trigger_key_positions and #ht.hold_trigger_key_positions > 0 then
            props["hold-trigger-key-positions"] = format_property(ht.hold_trigger_key_positions)
        end
        if ht.hold_trigger_on_release ~= nil then
            props["hold-trigger-on-release"] = format_property(ht.hold_trigger_on_release)
        end
        if ht.flavor then
            props["flavor"] = format_property(ht.flavor)
        end
        result[#result + 1] = {
            name = ht.name,
            description = ht.description or "",
            compatible = "zmk,behavior-hold-tap",
            binding_cells = 2,
            bindings = ht.bindings,
            properties = props,
        }
    end
    return result
end

local function build_standard_layout()
    return {
        layers = build_layers(),
        combos = combos_data.combos,
        behaviors = build_behaviors(),
        macros = build_macros(),
        input_listeners = combos_data.input_listeners,
        metadata = combos_data.metadata,
    }
end

local function save_keymap()
    local standard_layout = build_standard_layout()
    local json = encode_json(standard_layout)
    local output_path = "out/tailorkey_from_json.keymap"

    os.execute("mkdir -p out")
    layout:parse_json(json, "examples/moergo_glove80.j2")
    layout:save_dts(output_path)
    log(string.format("Recreated TailorKey layout -> %s", output_path))
end

save_keymap()
